Macro Hygiene + Locking Hygiene

jsuereth opened the pull request:
This PR does two things:
1. We migrate all instances of GRL locks into dynamic sections of pickling.  ANything which _generates_ a runtime pickler is locked.   All runtime picklers which use reflection lock in their pickle/unpickle methods.
2. We fully qualify ALL types in the macro in `_root_.` which means we can remove the imports from the generated macro trees.

Should take care to review, but all tests pass and I tried an experiment with overriding sbt-server using this code, but not sure I fully refreshed the generated macro code to be clear of GRL locks.

Review by> @eed3si9n @phaller

jsuereth committed:
Add a set of abstract classes to make creating picklers from Java easier.

jsuereth committed:
Remove runtime serialization locking from most pickling methods.

jsuereth committed:
More cleanups for hygiene issues.

jsuereth committed:
Fix remaining hygiene issues, evidenced by removing the imports.

jsuereth committed:
Finish pushing GRL locking until to dynamic pickling sections, rather than on EVERY pickle/unpickle call.

jsuereth commented:
A quick note on benchmarks:

I've spot checked a few, and we're definitely down on performance from what was recorded for the paper.   I'm ok with that as we can still beat Kryo serialization generally.

What I am sour about is that the _first_ pickle time (from zero configuration) is now astronomically high.  I think we may need to dig into why this is.  I suspect a few things:
1. Initialization time of various "static" caches and objects (we added a cake to bake)
2. Micro-locking that occurs now on a cache-miss looking up runtime picklers in these tests.
3. General un-optimisations we accrued during interface cleanup.

Something for us to investigate when we have time, but for now we need to continue cleaning up after the macros, and providing more stability for non-hand-generated picklers.

eed3si9n commented:
Do our tests run concurrently? Is there a set of tests that stress test the GRL aspect?

jsuereth commented:
I tested this against sbt-server which had caught the GRL issues in the past.

phaller commented:
I'll have a look at it. (Currently in transit, so delays possible.)

phaller committed:
Better documentation, clean-ups

- Add GRL locking/unlocking when creating runtime pickler
  in `PicklerMacros`.
- Clarify API of trait `Pickle`.
- Minor cosmetic fixes.

jsuereth commented:
Thanks for the additionally fixes @phaller !  It's looking great to me.

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -48,18 +48,14 @@ trait PicklerUnpicklerMacros extends Macro
-        override def pickle(picklee: $tpe, builder: scala.pickling.PBuilder): Unit = $picklerTree
-        override def unpickle(tagKey: String, reader: scala.pickling.PReader): Any = $unpicklerTree
-        override def tag: FastTypeTag[$tpe] = $createTagTree
+        override def pickle(picklee: $tpe, builder: _root_.scala.pickling.PBuilder): Unit = $picklerTree
```
`Unit` is not fully qualified

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -48,18 +48,14 @@ trait PicklerUnpicklerMacros extends Macro
-        override def unpickle(tagKey: String, reader: scala.pickling.PReader): Any = $unpicklerTree
-        override def tag: FastTypeTag[$tpe] = $createTagTree
+        override def pickle(picklee: $tpe, builder: _root_.scala.pickling.PBuilder): Unit = $picklerTree
+        override def unpickle(tagKey: String, reader: _root_.scala.pickling.PReader): Any = $unpicklerTree
```
`Any` is not fully qualified

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -344,12 +341,11 @@ trait PicklerMacros extends Macro with PickleMacros with FastTypeTagMacros {
-          import _root_.scala.pickling.PickleOps
-          def pickle(picklee: $tpe, builder: scala.pickling.PBuilder): Unit = $pickleLogicTree
-          def tag: FastTypeTag[$tpe] = $createTagTree
+          def pickle(picklee: $tpe, builder: _root_.scala.pickling.PBuilder): Unit = $pickleLogicTree
```
`Unit` is not fully qualified

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -362,11 +358,10 @@ trait PicklerMacros extends Macro with PickleMacros with FastTypeTagMacros {
         import _root_.scala.pickling.internal._
-        import _root_.scala.pickling.PickleOps
-        def pickle(picklee0: $tpe, builder: scala.pickling.PBuilder): Unit = $dpicklerPickleImpl
+        def pickle(picklee0: $tpe, builder: _root_.scala.pickling.PBuilder): Unit = $dpicklerPickleImpl
```
`Unit` is not fully qualified

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -427,9 +422,9 @@ trait OpenSumUnpicklerMacro extends Macro with UnpicklerMacros with FastTypeTagM
-        def unpickle(tagKey: String, reader: scala.pickling.PReader): Any = $unpickleLogic
-        def tag: scala.pickling.FastTypeTag[$tpe] = $createTagTree
+      implicit object $unpicklerName extends _root_.scala.pickling.Unpickler[$tpe] with _root_.scala.pickling.Generated {
+        def unpickle(tagKey: String, reader: _root_.scala.pickling.PReader): Any = $unpickleLogic
```
`Any` is not fully qualified

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -541,20 +536,21 @@ trait UnpicklerMacros extends Macro with UnpickleMacros with FastTypeTagMacros {
                 val methodName = "set" + name
                 // obtain Class of parameter
                 val className = fir.tpe.toString
+                // TODO - Which of these are hygiene issues?
```
``` scala
scala> def test(tp: Type) = if (tp.typeSymbol == definitions.ByteClass) "byte" else "other"
test: (tp: reflect.runtime.universe.Type)String

scala> test(typeOf[Short])
res0: String = other

scala> test(typeOf[Byte])
res1: String = byte
```

jsuereth replied:
I'm not sure what this means....

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -697,17 +692,16 @@ trait PickleMacros extends Macro with TypeAnalysis {
 
   def createPickler(tpe: c.Type, builder: c.Tree): c.Tree = q"""
-    val pickler = implicitly[scala.pickling.Pickler[$tpe]]
+    val pickler = implicitly[_root_.scala.pickling.Pickler[$tpe]]
```
`implicitly` is not fully qualified.

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -795,24 +785,16 @@ trait UnpickleMacros extends Macro with TypeAnalysis {
-        scala.pickling.internal.GRL.unlock()
-      }
+      var $unpicklerName: _root_.scala.pickling.Unpickler[$tpe] = null
+      $unpicklerName = implicitly[_root_.scala.pickling.Unpickler[$tpe]]
```
`implicitly` is not fully qualified

jsuereth replied:
Thanks for the hygiene point-out @retronym, very helpful!

retronym commented on core/src/main/scala/scala/pickling/PBuilderReader.scala:
```diff
@@ -104,6 +104,9 @@ trait PBuilder extends Hintable {
   def result(): Pickle
 }
+/** Abstract shim for Java. */
+abstract class AbtractPBuilder extends PBuilder with PickleTools
```
typo

retronym commented on core/src/main/scala/scala/pickling/Pickler.scala:
```diff
@@ -83,15 +84,18 @@ trait Unpickler[T] {
   /** The fast type tag associated with this unpickler. */
   def tag: FastTypeTag[T]
 }
+/** Shim for Java code. */
```
Doesn't seem useful as a Scaladoc comment, should this instead be an implementation comment?

retronym commented on core/src/main/scala/scala/pickling/pickler/List.scala:
```diff
@@ -29,7 +29,7 @@ trait CollectionPicklerUnpicklerMacro extends Macro with UnpickleMacros {
     val picklerUnpicklerName = c.fresh(syntheticPicklerUnpicklerName(tpe).toTermName)
     q"""
-      implicit object $picklerUnpicklerName extends scala.pickling.Pickler[$tpe] with scala.pickling.Unpickler[$tpe] {
+      implicit object $picklerUnpicklerName extends scala.pickling.AbstractPicklerUnpickler[$tpe] {
```
Missing `_root_` here. Code below seems to have missed the hygenification pass.

jsuereth replied:
I think this code is not actually used any longer...  Let me check though.

retronym commented on core/src/main/scala/scala/pickling/runtime/RuntimePickler.scala:
```diff
@@ -48,6 +48,7 @@ class RuntimeTypeInfo(classLoader: ClassLoader, clazz: Class[_], share: refs.Sha
   def shouldBotherAboutLooping(tpe: Type) = shareAnalyzer.shouldBotherAboutLooping(tpe)
 }
 
+// Note: This entire class must be constructed inside of a GRL-locked method.
```
Could this comment be made into an assertion?

jsuereth replied:
interesting, yeah probably.

retronym commented:
AFAIU, the GRL is used to protect two thread-unsafe entities: the Scala 2.10.x implementation of runtime reflection, and the global mutable state in the pickling library itself (caches of picklers generated at runtime).

It would be desirable to use separate locks for these to make it easier to remove the synchronization around Scala 2.11.x reflection. I would hope this could flush out bugs in the underlying reflection implementation itself.

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -504,7 +499,7 @@ trait UnpicklerMacros extends Macro with UnpickleMacros with FastTypeTagMacros {
           q"${sym.module}"
         } else if (cir.javaGetInstance) {
-          q"""java.lang.Class.forName(${tpe.toString}).getDeclaredMethod("getInstance").invoke(null)"""
+          q"""_root_.java.lang.Class.forName(${tpe.toString}).getDeclaredMethod("getInstance").invoke(null)"""
```
`tpe.toString` is extremely fragile here. Suggestion:

```
scala> val tp = typeOf[java.util.Arrays]
tp: reflect.runtime.universe.Type = java.util.Arrays

scala> tp.typeSymbol.fullName
res5: String = java.util.Arrays

scala> Class.forName(res5)
res6: Class[_] = class java.util.Arrays
```

retronym replied:
I have know idea why you would generate code to perform Java reflection like this. What prevents us from generating the call to `$tpe.getInstance` directly?

jsuereth replied:
I think it's just a matter of cowboy coding something fast.

FYI - I plan to heavily rework the macros in the coming months (assuming I have free time).

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -627,10 +623,10 @@ trait UnpicklerMacros extends Macro with UnpickleMacros with FastTypeTagMacros {
-            } else if (tagKey == scala.pickling.FastTypeTag.Ref.key) {
-              implicitly[Unpickler[refs.Ref]].unpickle(tagKey, reader)
+            } else if (tagKey == _root_.scala.pickling.FastTypeTag.Ref.key) {
+              implicitly[_root_.scala.pickling.Unpickler[_root_.scala.pickling.refs.Ref]].unpickle(tagKey, reader)
```
Here's another unqualfied `implicitly`.

retronym commented on core/src/main/scala/scala/pickling/Macros.scala:
```diff
@@ -763,7 +753,7 @@ trait UnpickleMacros extends Macro with TypeAnalysis {
 
   def createUnpickler(tpe: Type): Tree =
-    q"implicitly[scala.pickling.Unpickler[$tpe]]"
+    q"implicitly[_root_.scala.pickling.Unpickler[$tpe]]"
```
unqualfied `implicitly`.

jsuereth committed:
Fix remaining hygiene issues in mainline macros, and formalize a working test for both scala 2.11 and 2.10.

jsuereth committed:
Add sequence hygienic tests, and various other cleanups from @retronym.

jsuereth commented:
Ok, @phaller @retronym @eed3si9n ,  This one is ready for a second look.

retronym commented:
LGTM.

FTR, an alternative to fully qualifying names is to splice symbols, 

``` scala
val FastTypeTagClass = typeOf[FastTypeTag[_]].typeSymbol // or just symbolOf[...] in 2.11
val IntTpe = typeOf[Int]
...

q"_root_.scala.pickling.FastTypeTag[_root_.scala.Int]" // Using FQNs
q"new $FastTypeTag[$IntTpe]]" // splicing symbols
```

phaller commented:
LGTM
